### PR TITLE
Cleanup ThreadContext in finally-clause

### LINKED ISSUE #292: Memory leak when logging exception telemetry
The `WebRequestTrackingFilter` uses `ThreadLocal` to store state. This state is not cleaned up in a `finally`-clause, but it's cleaned up on [this line](https://github.com/Microsoft/ApplicationInsights-Java/blob/master/web/src/main/java/com/microsoft/applicationinsights/web/internal/WebRequestTrackingFilter.java#L187). That line is never reached if `chain.doFilter` throws an exception since the exception handling rethrows the exception.

By the way, wouldn't the rethrow make it impossible to correlate the exception with the request? Since no telemetry in sent with request information? How does that correlation work?

By the way2: It's good practice to have classes not marked with `final` and use `protected` for fields and methods when writing libraries. When bugs such as these occur, it is much easier for clients to deal with them. With the code as it stands now, I'm left with forking the project and releasing an in-house version with a fix, or copy all the code into my code base.

### PR DESCRIPTION
**Pull request by jgogstad**
Fixes #292

### CODE REVIEW

**Comment by OriYosefiMSFT** [comment]
Thanks for contributing the solution. I understand the issue you are reporting but I'm concerned that the proposed code may have an issue as well. 
webModulesContainer.invokeOnEndRequest would invoke the module's OnEndRequest which uses the TLS (e.g. [here](https://github.com/Microsoft/ApplicationInsights-Java/blob/master/web/src/main/java/com/microsoft/applicationinsights/web/extensibility/modules/WebRequestTrackingTelemetryModule.java#L113) ).
After the proposed changed, the TLS would have already been cleaned up in the "finally" clause.

File: web/src/main/java/com/microsoft/applicationinsights/web/internal/WebRequestTrackingFilter.java [hunk 1]
```diff
@@ -181,10 +182,6 @@ private boolean invokeSafeOnBeginRequest(ServletRequest req, ServletResponse res
     private void invokeSafeOnEndRequest(ServletRequest req, ServletResponse res) {
         try {
             webModulesContainer.invokeOnEndRequest(req, res);
```
Comment (OriYosefiMSFT) [inline web/src/main/java/com/microsoft/applicationinsights/web/internal/WebRequestTrackingFilter.java]: This call invokes the modules' onEndRequest method which uses the TLS. After cleanup above, the TLS would no longer be available

**Comment by jgogstad** [comment]
True, sorry for the lack of proper testing on my part. Pushed a fix that also addresses #298. I restructured the conditionals a bit in order to make the exception handling more readable. As long as you're targeting JRE 1.6, we can't collapse the catch-blocks so the code is duplicated.

**Comment by gupele** [comment]
Thanks for the fix!

**Pull request merged**